app/actions/handlers.py

```diff
@@ -1,28 +1,194 @@
+from app.services.gundi import send_observations_to_gundi
+from app.services.state import IntegrationStateManager
 
-from .configurations import AuthenticateConfig, PullObservationsConfig
 
-For example, this "auth" action will authenticate against the third-party system to test credentials
-async def action_auth(integration, action_config):
-    # Custom logix to authenticate
-    # The action handler can return a result (optional). In this case it's the result of the connection test.
-    # ToDo: Standardize these responses
-    return {"valid_credentials": token is not None} 
+logger = logging.getLogger(__name__)
 
-This "pull_observations" action implements the data extraction
-async def action_pull_observations(integration, action_config):
-    # Authenticate ..
-    # Extract Data ..
-    # Transform it ..
-    # Push it to Gundi API v2 ..
-    # Return a result (optional)
-    return {"observations_extracted": 100}
 
+state_manager = IntegrationStateManager()
 
-You can add here any other helper functions as needed too. Just ensure that they don't start with the "action_" prefix
-or they will be considered actions.
 
-"""
\ No newline at end of file
+async def filter_and_transform(vehicles, integration_id, action_id):
+    transformed_data = []
+    main_data = ["deviceId", "name", "timeStr", "time", "y", "x"]
+    for vehicle in vehicles:
+        # Get current state for the device
+        current_state = await state_manager.get_state(
+            integration_id,
+            action_id,
+            vehicle.deviceId
+        )
+
+        if current_state:
+            # Compare current state with new data
+            latest_device_timestamp = datetime.datetime.strptime(
+                current_state.get("latest_device_timestamp"),
+                '%Y-%m-%d %H:%M:%S'
+            )
+
+            if vehicle.timeStr <= latest_device_timestamp:
+                # Data is not new, not transform
+                logger.info(
+                    f"Excluding device ID '{vehicle.deviceId}' obs '{vehicle.timeStr}'"
+                )
+                continue
+
+        data = {
+            "source": vehicle.deviceId,
+            "source_name": vehicle.name,
+            'type': 'tracking-device',
+            "recorded_at": vehicle.timeStr,
+            "location": {
+                "lat": vehicle.y,
+                "lon": vehicle.x
+            },
+            "additional": {
+                key: value for key, value in vehicle.dict().items()
+                if key not in main_data and value is not None
+            }
+        }
+        transformed_data.append(data)
+
+    return transformed_data
+
+
+async def action_auth(integration, action_config: AuthenticateConfig):
+    logger.info(f"Executing auth action with integration {integration} and action_config {action_config}...")
+    try:
+        token = await client.get_auth_token(
+            integration=integration,
+            config=client.get_auth_config(integration)
```

Comment:
```suggestion
            config=action_config
```
Same as the other comment, now you get what you need in `action_config` :)